File: packages/node/src/parse-args.ts
/** The CLI arguments the TypeScript mutation adapter understands. */
export interface AdapterArgs {
  /** Stryker `mutate` patterns (`--mutate a,b,...`), or `undefined` for the default set. */
  mutate?: string[];
  /** File to write the normalized-results JSON to (`--out <path>`); stdout when absent. */
  out?: string;
}

/**
 * Parse the adapter's CLI arguments: `--mutate <a,b,...>` (comma-separated Stryker mutate
 * patterns) and `--out <path>` (where to write the normalized JSON; the rule passes a temp
 * file so Stryker's own stdout logging never corrupts the results). A flag with no following
 * value is treated as absent.
 */
export function parseArgs(argv: string[]): AdapterArgs {
  const args: AdapterArgs = {};

  const mutateIdx = argv.indexOf('--mutate');
  const mutateValue = mutateIdx === -1 ? undefined : argv[mutateIdx + 1];
  if (mutateValue !== undefined) {
    args.mutate = mutateValue.split(',');
  }

  const outIdx = argv.indexOf('--out');
  const outValue = outIdx === -1 ? undefined : argv[outIdx + 1];
  if (outValue !== undefined) {
    args.out = outValue;
  }

  return args;
}

File: packages/python/python/testing_conventions/mutation/parse_args.py
"""Parse the mutation adapter's CLI arguments.

The rust binary spawns ``python -m testing_conventions.mutation.main`` and passes these:
``--out <path>`` (where to write the normalized-results JSON — the rule passes a temp file)
and ``--module <path>`` (repeatable; the source files to mutate, for a diff-scoped run).
Without any ``--module`` the whole project is mutated.
"""
from __future__ import annotations

import argparse
from dataclasses import dataclass, field


@dataclass
class Args:
    """The adapter's parsed arguments."""

    out: str
    modules: list[str] = field(default_factory=list)


def parse_args(argv):
    """Parse ``argv`` (a list of strings) into [`Args`]."""
    parser = argparse.ArgumentParser(prog="testing-conventions-mutation")
    parser.add_argument("--out", required=True)
    parser.add_argument("--module", action="append", default=[], dest="modules")
    ns = parser.parse_args(argv)
    return Args(out=ns.out, modules=ns.modules)
